Todo Model の作成

chionyan opened the pull request:
## 概要
### UUID の設定
- マイグレーション時に `pgcrypto` を作成する拡張の設定
- テーブルのマイグレーションファイルを作成した時にプライマリーキーを UUID にする設定
- 参考：[RailsのモデルIDにUUIDを使う（翻訳）](https://techracho.bpsinc.jp/hachi8833/2018_01_04/50565/)
- 参考：[Active Record and PostgreSQL 2 UUID Primary Keys](https://edgeguides.rubyonrails.org/active_record_postgresql.html#uuid-primary-keys)

### Todo Model の作成
- バリデーションの設定
  - `Todo#title`, `Todo#text` が空でないことのバリデーション設定
- DB のマイグレーション
  - マイグレーションファイルに NOT NULL 制約を追加
  - `$ rails db:migrate` で DB のマイグレーション
- spec ファイルの作成
  - `shoulda-matchers` の設定
  - Model Spec の作成

## 動作確認
- [x] Rails
  - `$ bin/setup` を実行
  - `$ bundle exec rails console` を実行
  - [x] 下記のように id に UUID が設定された Todo インスタンスが3件出力されることを確認
  ```ruby
  [1] pry(main)> Todo.all
    Todo Load (1.3ms)  SELECT "todos".* FROM "todos"
  => [#<Todo:0x00007fba0d0688c0 id: "f0903cf0-2231-4936-9947-5e0cdad46b39", title: "title_0", text: "text_0", created_at: Mon, 13 Aug 2018 06:36:15 UTC +00:00, updated_at: Mon, 13 Aug 2018 06:36:15 UTC +00:00>,
  #<Todo:0x00007fba0d05e9d8 id: "b6437798-aac5-4de3-8b4d-4e85372c02b1", title: "title_1", text: "text_1", created_at: Mon, 13 Aug 2018 06:36:15 UTC +00:00, updated_at: Mon, 13 Aug 2018 06:36:15 UTC +00:00>,
  #<Todo:0x00007fba0d05e870 id: "882b0ed3-0655-4984-876c-a2f3fa3d751c", title: "title_2", text: "text_2", created_at: Mon, 13 Aug 2018 06:36:15 UTC +00:00, updated_at: Mon, 13 Aug 2018 06:36:15 UTC +00:00>]
  ```

- [x] RSpec
  - `$ bundle exec rspec` を実行
  - [x] `0 failures` が出力されることを確認

- [x] Rubocop
  - `$ bundle exec rubocop` を実行  
  - [x] `no offenses detected` が出力されることを確認

chionyan committed:
Setting Using UUID

- https://edgeguides.rubyonrails.org/active_record_postgresql.html

chionyan committed:
$ rails g resource todo title:string text:string

chionyan committed:
$ rails db:migrate

- title, text に NOT NULL 制約を追加

chionyan committed:
Setting Using shoulda-matchers

chionyan committed:
Set validation Todo#title, Todo#text

chionyan committed:
Edit db/seeds.rb

chionyan commented:
@koshigoe 
Todo Model 作成のPRを作成しました！
よろしければレビューお願いします🙏✨

chionyan commented:
(口頭説明)

PostgreSQL拡張を有効にする `.enable_extension` は内部的に下記のようなクエリが実行されている
```ruby
def enable_extension(name)
  exec_query("CREATE EXTENSION IF NOT EXISTS \"#{name}\"").tap {
    reload_type_map
  }
end
```

参考：[enable\_extension \(ActiveRecord::ConnectionAdapters::PostgreSQLAdapter\) \- APIdock](https://apidock.com/rails/ActiveRecord/ConnectionAdapters/PostgreSQLAdapter/enable_extension)

koshigoe commented:
> You need to enable the pgcrypto (only PostgreSQL >= 9.4) or uuid-ossp extension to use uuid.
> https://edgeguides.rubyonrails.org/active_record_postgresql.html#uuid

🙄 

> 歴史的にこのモジュールは、モジュールの名前の由来となったOSSP UUIDライブラリに依存していました。 OSSP UUIDライブラリはまだhttp://www.ossp.org/pkg/lib/uuid/にありますが、あまりよく維持されておらず、より新しいプラットフォームへ移植することがますます困難になってきています。
> ...
> 注意: ランダムに生成された(バージョン4)UUIDのみが必要な場合には、代わりにpgcryptoモジュールのgen_random_uuid()を利用すること検討してください。
> https://www.postgresql.jp/document/9.4/html/uuid-ossp.html

🙄

把握。

koshigoe commented on db/migrate/20180813061707_create_todos.rb:
```diff
@@ -0,0 +1,10 @@
+      t.timestamps
+    end
+  end
+end
```
❓Swagger の API 仕様から読み取りきれなかったのですが、 `text` は `string` (character varying)で良いのですよね？

```
# \d todos
                             Table "public.todos"
   Column   |            Type             |             Modifiers
------------+-----------------------------+------------------------------------
 id         | uuid                        | not null default gen_random_uuid()
 title      | character varying           | not null
 text       | character varying           | not null
 created_at | timestamp without time zone | not null
 updated_at | timestamp without time zone | not null
Indexes:
    "todos_pkey" PRIMARY KEY, btree (id)
```

koshigoe replied:
❓ユニーク制約とかの話も、今は気にしなくて良いですよね？

chionyan replied:
こちらについて私も確認が取れておらず、 Swagger 下部の Models にある、
<img width="1200" alt="2018-08-13 16 44 42" src="https://user-images.githubusercontent.com/26681827/44018915-43b7e370-9f18-11e8-856f-657f393acfb4.png">
から、型は全て string であると想定して作成していました…！

また、[大道君のPR](https://github.com/daido1976/simple-todo-api/pull/3#discussion_r199049646)に

 > swgger の * マークが必須項目だった

と書かれているので、全て入力必須なのかなと思っています 🤔 
ユニーク制約については今の所確認できませんでした…！

※現在確認中のため、確認が取れ次第また連絡します❗️

koshigoe replied:
Swagger の型と PostgreSQL の型は異なるので、DB に文字列を格納する際に DB のどの型が適切かを確認できるとよりよいかと思いました。

が、制約の件含め、あとで調整という事でも構わない部分だとおもいます。

chionyan replied:
https://feedforce.slack.com/archives/CA0CDSLLW/p1534147578000291?thread_ts=1534146035.000184&cid=CA0CDSLLW
とのことで、ユニークは `id` だけで良いということでした！

> Swagger の型と PostgreSQL の型は異なる

知らなかったです😔 `uuid` `character varying` 共に文字列として格納するのですね…！

koshigoe replied:
🤔 アプリケーションが取り扱うデータ型について、フォローした方がよいかな？

- Swagger における型は API レスポンス上の表現(e.g. `string`)
- Rails Migration における型はデータベースの型を抽象化した表現(e.g. `string`, `text`)
- PostgreSQL における型は PostgreSQL 固有の表現(e.g. `character`, `character varying`, `text`)

👆上記の様な区別ができていれば問題ないです。 🙇

koshigoe commented on db/seeds.rb:
```diff
@@ -1,7 +1,3 @@
-#   Character.create(name: 'Luke', movie: movies.first)
+3.times do |num|
+  Todo.create(title: "title_#{num}", text: "text_#{num}")
+end
```
❓ ここは `create!` ではなく `create` で良いですよね？(あまり気にしなくて良いですよね？)

chionyan replied:
個人的に条件分けされていないので、 `create!` の方がいいのかな…と思いました💦

chionyan replied:
9f0aa4a で `create!` に修正しました！

chionyan commented:
`uuid-ossp` を入れてたので `DROP EXTENSION uuid-ossp;` してみましたが、上記の動作確認できました！ 公式が `pgcrypto` を推奨するならそちらで拡張した方が良さそうですね！

koshigoe reviewed (commented):
いくつか仕様的に気になった点はありつつ、PR としてはおおむね問題ない様に思う。

- コントローラやシリアライザも(テストなしで)作っている点は若干気になるけど、気にしすぎなくて良い気がしている
- ルーティングも CRUD 全対応で追加されている点は若干気になるけど、気にしすぎなくて良い気がしている

※ 個人的には、master に向けた PR は必要最低限の粒度であるべきだと考えているので、未完のルーティングやコントローラも含めるのはスコープとして広すぎる様な印象がある。
※ が、初速を出すという意味ではあまり気にしすぎなくても良い気はする。

koshigoe commented:
中途半端なレビューになってしまったな。。。

タイトル通りにモデル作成が主題なら、実際に実装している内容の範囲が広すぎるので、必要最低限(モデルに関する内容)に絞るべき。

ある程度、ジェネレータで生成した半端な実装を許容するポリシーなら、このままマージしてもよいと思う。

koshigoe commented:
> Todo Resource (Model, Controller, Serializer, Routing) の作成

上記に関するこのPR でのゴールを教えてもらえると助かります。

この PR によって、このアプリケーションは何ができるようになりますか？

概要として細かな作業内容をまとめてくれることはたすかりますが、達成すべきゴールについても、漏れなく簡潔に書いてもらえると大変助かります。 :pray:

chionyan commented:
この PR における目的はタイトルの通り、「Todo Model の作成」のみとするため、 Model の実装のみのPRとします！
したがって、
`$ rails g resource todo title:string text:string` で Controller, Serializer, Routing を作成していたところを、 `$ rails g model todo title:string text:string` で Model と DB に関する必要最低限のファイルだけの作成を目的とします！

chionyan committed:
Delete Todor Resource Files except Model, Migrate File

- $ rails g model todo title:string text:string の状態になるように不要なファイルを削除
- マイグレーションファイルを作成しているため、コミットのハッシュ値が変わるため、手動で削除

chionyan commented:
068f3aa で a60e1f8 が `$ rails g model todo title:string text:string` コマンドだった時を模した形を再現しました！
手動で行なった理由としては、DB の migrate の関係で作成されるファイル名などが変わってしまうためか、コミットのハッシュ値が変わってしまうため、既にレビュー依頼を行なっているので手動で該当ファイルを削除するコミットを作成しました🙇‍♀️

chionyan committed:
Change Todo.create -> Todo.create!

koshigoe reviewed (approved):
よさげ。

LGTM
